vscode/src/azure/networkRequests.ts

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { log } from "qsharp-lang";
+import { workspace } from "vscode";
+
+const publicMgmtEndpoint = "https://management.azure.com";
+
+export async function azureRequest(
+  uri: string,
+  token: string,
+  method = "GET",
+  body?: string
+) {
+  const headers: [string, string][] = [
+    ["Authorization", `Bearer ${token}`],
+    ["Content-Type", "application/json"],
+  ];
+
+  try {
+    log.debug(`Fetching ${uri} with method ${method}`);
+    const response = await fetch(uri, {
+      headers,
+      method,
+      body,
+    });
+
+    if (!response.ok) {
+      log.error("Azure request failed", response);
+      throw Error(`Azure request failed: ${response.statusText}`);
+    }
+
+    log.debug(`Got response ${response.status} ${response.statusText}`);
+    const result = await response.json();
+    log.trace("Response value: ", result);
+
+    return result;
+  } catch (e) {
+    log.error(`Failed to fetch ${uri}: ${e}`);
+    throw e;
+  }
+}
+
+// Different enough to above to warrant it's own function
+export async function storageRequest(
+  uri: string,
+  method: string,
+  extraHeaders?: [string, string][],
+  body?: string | Uint8Array
+) {
+  const headers: [string, string][] = [
+    ["x-ms-version", "2023-01-03"],
+    ["x-ms-date", new Date().toUTCString()],
+  ];
+  const storageProxy: string | undefined = workspace
+    .getConfiguration("Q#")
+    .get("storageProxy"); // e.g. in settings.json: "Q#.storageProxy": "https://qsx-proxy.azurewebsites.net/api/proxy";
+
+  if (extraHeaders?.length) headers.push(...extraHeaders);
+  if (storageProxy) {
+    log.debug(`Setting x-proxy-to header to ${uri}`);
+    headers.push(["x-proxy-to", uri]);
+    uri = storageProxy;
+  }
+  try {
+    log.debug(`Fetching ${uri} with method ${method}`);
+    const response = await fetch(uri, { method, headers, body });
+    if (!response.ok) {
+      log.error("Storage request failed", response);
+      throw Error(`Storage request failed: ${response.statusText}`);
+    }
+    log.debug(`Got response ${response.status} ${response.statusText}`);
+    return response;
+  } catch (e) {
+    log.error(`Failed to fetch ${uri}: ${e}`);
+    throw e;
+  }
+}
+
+export class AzureUris {
```

Comment:
Wouldn't it be easier to use our JavaScript client library for this? https://learn.microsoft.com/en-us/javascript/api/overview/azure/quantum-jobs-readme?view=azure-node-preview

I guess I wouldn't be surprised to learn that one doesn't work in the browser, or something...

Reply:
I spend hours digging through all the various JavaScript libraries for Azure authentication, ARM subscriptions, blob storage APIs, quantum requests, etc. They all were problematic is some ways, such as not working in Web Workers, using now deprecated libraries, having dependencies on other VS Code extensions being installed, etc.

Down the road we could look at adopting a dependency again to make this easier, but right now this was a confusing and problematic mess.